capsules/src/i2c_master_slave_driver.rs

```diff
@@ -222,73 +237,133 @@ impl hil::i2c::I2CHwSlaveClient for I2CMasterSlaveDriver<'_> {
 impl Driver for I2CMasterSlaveDriver<'_> {
     fn allow_readwrite(
         &self,
-        _appid: ProcessId,
+        process_id: ProcessId,
         allow_num: usize,
         mut slice: ReadWriteAppSlice,
     ) -> Result<ReadWriteAppSlice, (ReadWriteAppSlice, ErrorCode)> {
-        match allow_num {
-            // Pass in a buffer for transmitting a `write` to another
-            // I2C device.
-            0 => {
-                self.app.map(|app| {
-                    core::mem::swap(&mut app.master_tx_buffer, &mut slice);
-                });
-                Ok(slice)
-            }
-            // Pass in a buffer for doing a read from another I2C device.
-            1 => {
-                self.app.map(|app| {
-                    core::mem::swap(&mut app.master_rx_buffer, &mut slice);
-                });
-                Ok(slice)
-            }
-            // Pass in a buffer for handling a read issued by another I2C master.
-            2 => {
-                self.app.map(|app| {
-                    core::mem::swap(&mut app.slave_tx_buffer, &mut slice);
-                });
-                Ok(slice)
-            }
-            // Pass in a buffer for handling a write issued by another I2C master.
-            3 => {
-                self.app.map(|app| {
-                    core::mem::swap(&mut app.slave_rx_buffer, &mut slice);
-                });
-                Ok(slice)
-            }
-            _ => Err((slice, ErrorCode::NOSUPPORT)),
+        // Check if this non-virtualized driver is already in use by
+        // some (alive) process
+        let match_or_empty_or_nonexistant = self.app.map_or(true, |current_process| {
+            self.apps
+                .enter(*current_process, |_| current_process == &process_id)
+                .unwrap_or(true)
+        });
+        if match_or_empty_or_nonexistant {
+            self.app.set(process_id);
+        } else {
+            return Err((slice, ErrorCode::NOMEM));
+        }
+        let app = self.app.expect("cant fail");
```

Comment:
Isn't `app` always equivalent to `process_id` here, given the logic above? Then I suppose we could just use that and get rid of the `expect`.

Reply:
good call, done

Reply:
Did you push this already?